File: FE/DATN-Fashion-Shop-FE/src/app/dto/Response/page-response.ts
export class PageResponse<T> {
  content: T[]; // Danh sách dữ liệu
  pageNo: number; // Số trang hiện tại
  pageSize: number; // Số phần tử mỗi trang
  totalPages: number; // Tổng số trang
  totalElements: number; // Tổng số phần tử
  first: boolean; // Là trang đầu tiên?
  last: boolean; // Là trang cuối cùng?

  constructor(data?: Partial<PageResponse<T>>) {
    this.content = data?.content || [];
    this.pageNo = data?.pageNo ?? 0;
    this.pageSize = data?.pageSize ?? 10;
    this.totalPages = data?.totalPages ?? 0;
    this.totalElements = data?.totalElements ?? 0;
    this.first = data?.first ?? false;
    this.last = data?.last ?? false;
  }
}

File: DATN-Fashion-Shop-BE/src/main/java/com/example/DATN_Fashion_Shop_BE/dto/response/PageResponse.java
package com.example.DATN_Fashion_Shop_BE.dto.response;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.*;
import org.springframework.data.domain.Page;

import java.util.List;

@Data
@Builder
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class PageResponse<T> {
    private List<T> content;  // Danh sách dữ liệu
    private int pageNo;       // Số trang hiện tại
    private int pageSize;     // Số phần tử mỗi trang
    private int totalPages;   // Tổng số trang
    private long totalElements; // Tổng số phần tử
    private boolean first;     // Là trang đầu tiên?
    private boolean last;       // Là trang cuối cùng?

    public static <T> PageResponse<T> fromPage(Page<T> page) {
        return PageResponse.<T>builder()
                .content(page.getContent())
                .pageNo(page.getNumber())
                .pageSize(page.getSize())
                .totalPages(page.getTotalPages())
                .totalElements(page.getTotalElements())
                .first(page.isFirst())
                .last(page.isLast())
                .build();
    }
}
